File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def findMid(self, head):
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        #notice that the first half doesnt change their order but the second half changes
        #find the middle node first
        mid_node = self.findMid(head)

        #then, we need to reverse the last half
        def reverse(head):
            prev = None
            cur = head

            while cur:
                nxt = cur.next
                cur.next = prev
                prev = cur
                cur = nxt
            return prev
        reverse_head = reverse(mid_node.next)
        mid_node.next = None #need to cut this, otherwise, it will form the cycle
        cur1= head
        cur2 = reverse_head
        while cur1 and cur2:
            nxt1 = cur1.next
            nxt2 = cur2.next
            cur1.next = cur2
            cur2.next= nxt1
            cur1 = nxt1
            cur2 = nxt2
        # return head

File: Data Structures & Algorithms/intersection-of-two-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        int l1 = 0;
        int l2 = 0;
        ListNode* p1 = headA;
        ListNode* p2 = headB;
        while(p1){
            p1 = p1->next;
            l1 ++;
        }
        while(p2){
            p2 = p2->next;
            l2 ++;
        }
        p1 = headA;
        p2 = headB;

        if (l2 > l1){
            int diff = l2 -l1;
            while(diff > 0){
                p2 = p2->next;
                diff--;
            }
        }else if (l2 < l1){
            int diff = l1 -l2;
            while(diff > 0){
                p1 = p1->next;
                diff --;
            } 
        }

        while(p1 and p2){
            if (p1 == p2){
                return p1;
            }
            p1 = p1->next;
            p2 = p2->next;
        }
        return nullptr;
    }
};